trello-history/trello-history.md

```diff
@@ -0,0 +1,212 @@
+testing purpose. So all we had to do was to create a new project from the
+[power up skeleton project](https://glitch.com/edit/#!/trello-power-up-skeleton)
+gracefully proposed by Trello and remix it to create our own service.
+
+Here is file structure of the project:
+```
+- views
+- public
+  |_ css
+  |_ image
+  |_ js
+  |_ translations
+- server.js
+```
+
+Nothing fancy here. The entry point of the application is served by an express
+instance. From here, we perform an authorization check, then render the history
+of the card if the user is authenticated, or a button to ask authorization if
+he's not:
+
+```js
+window.TrelloPowerUp.initialize({
+  'card-back-section': (t, options) => getTranslations(locale)
+    .then(response => response.json())
+    .then(trans => t.set('organization', 'shared', 'trans', trans))
+    .then(() => t.getRestApi()
+      .isAuthorized()
+      .then(isAuthorized => isAuthorized
+        ? renderHistory(t)
+        : askAuthorization(t)
+      ),
+    )
+}, {
+  appKey: 'no-you-wont-have-my-key',
+  appName: 'KNP Trello History',
+})
+```
+
+Some explanations on this piece of code: The `initialize` method takes an object
+as first parameter, which associates each capability with the callback it should
+execute (in this case, as soon as the back of a card is revealed).
+
+Keep in mind that power up capabilities are rendered in iframes. Yes. Please
+don't leave. The `renderHistory` and `askAuthorization` at the end of the
+promise flow are simple functions that both returns descriptive objects of how
+these iframes should be displayed. The important thing to understand is here:
+Each iframe targets a specific HTML page which runs its own JS script. That is,
+how can we possibily **share data** between those scripts ?
+
+See that `t` argument ? It's the power up toolbox, exposing helpful methods to
+interact with the context of the Trello board. For example, the `t.set()` method
+is to be used to store scalar values in the context (translations, for example,
+are a perfect use case for it). Once stored, a `t.get()` method can be called
+from any other script to get our translations back. Piece of cake !
+
+This can be missleading, though. Tricked by our daily habits, we spent some time
+trying to over decouple things (i.e get the history data from the API in the
+main power up script above, store it in the context and only use other scripts
+for templating). The Trello buffer has a limit size of 4096 characters, so no
+need to say we didn't have any chance to successfullly store our stringified
+history there ! We wandered towards locale storage and string compression to
+bypass that limitation but finally figured out this is not how it's was supposed
+to be done.
+
+One script per iframe / page to be shown, simple as that. Let's have a quick
+look on the history script:
+
+```js
+t.render(() => t.get('organization', 'shared', 'translations')
+  .then(renderHistory(t))
```

Comment:
`renderHistory(translations)` otherwise we can mingle with the t from the power up

Reply:
`renderHistory` is a curried function, so we don't "see" the "translations" it takes as a second parameter.

Reply:
:ah: !